Add documentation for how to run jobs with Flux (#990)
fluxcd/website: content/en/docs/use-cases/running-jobs.md

---
title: Running pre and post-deployment jobs with Flux
linkTitle: Running Kubernetes Jobs with Flux
description: "How to run Kubernetes Jobs before and after an application deployment with Flux."

> weight: 10

Reviewer: Not sure but I think this value determines the order of the page at https://deploy-preview-990--fluxcd.netlify.app/docs/use-cases/ and I'm not sure how the ordering is supposed to look like. @stefanprodan might know. Maybe we just don't care.

Reviewer 2: I will fix the ordering and titles of all guides under use-cases in a a different PR. Thanks for bringing it up, I wanted to to this long time ago and I forgot about it.

---

> Additional considerations have to be made when managing Kubernetes Jobs with Flux.

Reviewer 3: Maybe make that a `note`.

By default, if you were to have Flux reconcile a Job resource,
it would apply it once to the cluster, the Job would create a Pod that can either error or run to completion.

> Attempting to update the Job manifest after it has been applied to the cluster is not be allowed, as changes to the

Reviewer 3: ```suggestion
Attempting to update the Job manifest after it has been applied to the cluster will not be allowed, as changes to the
```

> Jobs `spec.Completions`, `spec.Selector` and `spec.Template` are not permitted by the Kubernetes API.

Reviewer 3: ```suggestion
Job `spec.Completions`, `spec.Selector` and `spec.Template` are not permitted by the Kubernetes API.
```

To be able to update a Kubernetes Job, the Job has to be recreated by first being

> removed and the reapplied to the cluster.

Reviewer: ```suggestion
removed and then reapplied to the cluster.
```

## Repository structure

A typical use case for running Kubernetes Jobs with Flux is to implement pre-deployment tasks

> for e.g. database scheme migration and post-deployment jobs e.g. cache refresh.

Reviewer: ```suggestion
for e.g. database scheme migration and post-deployment jobs (like cache refresh).
```
omits the repetition of "e.g." that's already used some words prior to this location.

This requires separate [Flux Kustomization](../components/kustomize/kustomization.md) resources
that depend on each other: one for running the pre-deployment Jobs,
one to deploy the application, and a 3rd one for running the post-deployment Jobs.

> Example of an application config repository:

Reviewer 3: ```suggestion
Example of an application configuration repository:
```

```

And a Flux Kustomization that reconciles it at `./flux/pre-deploy.yaml`:

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: app-pre-deploy
spec:
  sourceRef:
    kind: GitRepository
    name: my-app
  path: "./pre-deploy/"
  interval: 60m
  timeout: 5m
  prune: true
  wait: true
  force: true
```

Setting `spec.force` to `true` will make Flux recreate the Job when any immutable field is changed,
forcing the Job to run every time the container image tag changes.
Setting `spec.wait` to `true` makes Flux wait for the Job to complete
before it is considered ready.

To deploy the application after the migration job,
we define a Flux Kustomization that depends on the migration one.

Example of `./flux/deploy.yaml`:

```yaml
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: app-deploy
spec:
  dependsOn:
    - name: app-pre-deploy
  sourceRef:
    kind: GitRepository
    name: my-app
  path: "./deploy/"
  interval: 60m
  timeout: 5m
  prune: true
  wait: true
```

This means that the `app-deploy` Kustomization will wait until all the Jobs in `app-pre-deploy` run to completion.

> If the Jobs fail, the app changes will not be applied by the `app-deploy` Kustomization.

Reviewer 3: ```suggestion
If the Job fails, the app changes will not be applied by the `app-deploy` Kustomization.
```
